### PR TITLE
next-upgrade: Ensure highest versions are used not latest

### PR DESCRIPTION
**Pull request by eps1lon**
`npm view <package>@<query>` returns packages in publish order. But we want the highest.

### CODE REVIEW

**Comment by eps1lon** [comment]
* **#73566** <a href="https://app.graphite.dev/github/pr/vercel/next.js/73566?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈 <a href="https://app.graphite.dev/github/pr/vercel/next.js/73566?utm_source=stack-comment-view-in-graphite" target="_blank">(View in Graphite)</a>
* `canary`

<h2></h2>

This stack of pull requests is managed by <a href="https://graphite.dev?utm-source=stack-comment"><b>Graphite</b></a>. Learn more about <a href="https://stacking.dev/?utm_source=stack-comment">stacking</a>.

File: packages/next-codemod/bin/upgrade.ts [hunk 1]
```diff
@@ -53,6 +53,8 @@ async function loadHighestNPMVersionMatching(query: string) {
   // npm-view returns an array if there are multiple versions matching the query.
   if (Array.isArray(versionOrVersions)) {
     // The last entry will be the latest version published.
```
Comment (eps1lon) [inline packages/next-codemod/bin/upgrade.ts:55]: I even spelled out the bug.

**Review by huozhi — APPROVED** [review]

**Comment by ijjk** [comment]
# Stats from current PR

<details open>
<summary><strong>Default Build</strong> (Increase detected ⚠️)</summary>

<br/>

<details>
<summary><strong>General</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| buildDuration | 19.4s | 17.4s | N/A |
| buildDurationCached | 16.6s | 14.4s | N/A |
| nodeModulesSize | 409 MB | 409 MB | N/A |
| nextStartRea..uration (ms) | 474ms | 480ms | N/A |

</details>

<details>
<summary><strong>Client Bundles (main, webpack)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| 1187-HASH.js gzip | 49.2 kB | 49.2 kB | N/A |
| 8276.HASH.js gzip | 169 B | 168 B | N/A |
| 8377-HASH.js gzip | 5.3 kB | 5.3 kB | N/A |
| bccd1874-HASH.js gzip | 53 kB | 53 kB | N/A |
| framework-HASH.js gzip | 57.5 kB | 57.5 kB | N/A |
| main-app-HASH.js gzip | 233 B | 235 B | N/A |
| main-HASH.js gzip | 33.7 kB | 33.7 kB | N/A |
| webpack-HASH.js gzip | 1.71 kB | 1.71 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Legacy Client Bundles (polyfills)</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| polyfills-HASH.js gzip | 39.4 kB | 39.4 kB | ✓ |
| Overall change | 39.4 kB | 39.4 kB | ✓ |

</details>

<details>
<summary><strong>Client Pages</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| _app-HASH.js gzip | 193 B | 193 B | ✓ |
| _error-HASH.js gzip | 193 B | 193 B | ✓ |
| amp-HASH.js gzip | 513 B | 511 B | N/A |
| css-HASH.js gzip | 343 B | 342 B | N/A |
| dynamic-HASH.js gzip | 1.84 kB | 1.84 kB | N/A |
| edge-ssr-HASH.js gzip | 265 B | 265 B | ✓ |
| head-HASH.js gzip | 363 B | 362 B | N/A |
| hooks-HASH.js gzip | 393 B | 392 B | N/A |
| image-HASH.js gzip | 4.44 kB | 4.43 kB | N/A |
| index-HASH.js gzip | 268 B | 268 B | ✓ |
| link-HASH.js gzip | 2.35 kB | 2.34 kB | N/A |
| routerDirect..HASH.js gzip | 328 B | 328 B | ✓ |
| script-HASH.js gzip | 397 B | 397 B | ✓ |
| withRouter-HASH.js gzip | 323 B | 326 B | N/A |
| 1afbb74e6ecf..834.css gzip | 106 B | 106 B | ✓ |
| Overall change | 1.75 kB | 1.75 kB | ✓ |

</details>

<details>
<summary><strong>Client Build Manifests</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| _buildManifest.js gzip | 746 B | 747 B | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Rendered Page Sizes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| index.html gzip | 524 B | 524 B | ✓ |
| link.html gzip | 538 B | 536 B | N/A |
| withRouter.html gzip | 520 B | 521 B | N/A |
| Overall change | 524 B | 524 B | ✓ |

</details>

<details>
<summary><strong>Edge SSR bundle Size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| edge-ssr.js gzip | 128 kB | 128 kB | N/A |
| page.js gzip | 200 kB | 200 kB | N/A |
| Overall change | 0 B | 0 B | ✓ |

</details>

<details>
<summary><strong>Middleware size</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| middleware-b..fest.js gzip | 667 B | 664 B | N/A |
| middleware-r..fest.js gzip | 155 B | 156 B | N/A |
| middleware.js gzip | 31.1 kB | 31.1 kB | N/A |
| edge-runtime..pack.js gzip | 844 B | 844 B | ✓ |
| Overall change | 844 B | 844 B | ✓ |

</details>

<details>
<summary><strong>Next Runtimes</strong></summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| 797-experime...dev.js gzip | 322 B | 322 B | ✓ |
| 797.runtime.dev.js gzip | 314 B | 314 B | ✓ |
| app-page-exp...dev.js gzip | 321 kB | 321 kB | ✓ |
| app-page-exp..prod.js gzip | 126 kB | 126 kB | ✓ |
| app-page-tur..prod.js gzip | 139 kB | 139 kB | ✓ |
| app-page-tur..prod.js gzip | 134 kB | 134 kB | ✓ |
| app-page.run...dev.js gzip | 311 kB | 311 kB | ✓ |
| app-page.run..prod.js gzip | 122 kB | 122 kB | ✓ |
| app-route-ex...dev.js gzip | 36.8 kB | 36.8 kB | ✓ |
| app-route-ex..prod.js gzip | 25 kB | 25 kB | ✓ |
| app-route-tu..prod.js gzip | 25 kB | 25 kB | ✓ |
| app-route-tu..prod.js gzip | 24.8 kB | 24.8 kB | ✓ |
| app-route.ru...dev.js gzip | 38.5 kB | 38.5 kB | ✓ |
| app-route.ru..prod.js gzip | 24.8 kB | 24.8 kB | ✓ |
| pages-api-tu..prod.js gzip | 9.56 kB | 9.56 kB | ✓ |
| pages-api.ru...dev.js gzip | 11.4 kB | 11.4 kB | ✓ |
| pages-api.ru..prod.js gzip | 9.56 kB | 9.56 kB | ✓ |
| pages-turbo...prod.js gzip | 21.3 kB | 21.3 kB | ✓ |
| pages.runtim...dev.js gzip | 27 kB | 27 kB | ✓ |
| pages.runtim..prod.js gzip | 21.3 kB | 21.3 kB | ✓ |
| server.runti..prod.js gzip | 916 kB | 916 kB | ✓ |
| Overall change | 2.34 MB | 2.34 MB | ✓ |

</details>

<details>
<summary><strong>build cache</strong> Overall increase ⚠️</summary>

|  | vercel/next.js canary  | vercel/next.js sebbie/12-05-next-upgrade_ensure_highest_versions_are_used_not_latest | Change |
| - | - | - | - |
| 0.pack gzip | 2.03 MB | 2.04 MB | ⚠️ +682 B |
| index.pack gzip | 146 kB | 144 kB | N/A |
| Overall change | 2.03 MB | 2.04 MB | ⚠️ +682 B |

</details>

<details>
<summary><strong>Diff details</strong></summary>

<details>
<summary>Diff for <strong>main-HASH.js</strong></summary>

Diff too large to display
</details>

</details>

</details>
Commit: 0cdf8f91dfb8c910372502ce6aa1b3a83c7e8539

**Comment by ijjk** [comment]
## Failing test suites 
Commit: 5bd64d548250e204267e03fe7a674065f69fa00b

`pnpm test test/integration/next-image-legacy/default/test/index.test.ts`
- Image Component Tests > development mode > should emit image for next/dynamic with non ssr case

<details>
<summary>Expand output</summary>

  ● Image Component Tests › development mode › should emit image for next/dynamic with non ssr case

    thrown: "Exceeded timeout of 60000 ms for a test.
    Add a timeout value to this test to increase the timeout, if this is a long-running test. See https://jestjs.io/docs/api#testname-fn-timeout."

      1168 |   })
      1169 |
    > 1170 |   it('should emit image for next/dynamic with non ssr case', async () => {
           |   ^
      1171 |     let browser = await webdriver(appPort, '/dynamic-static-img')
      1172 |     const img = await browser.elementById('dynamic-loaded-static-jpg')
      1173 |     const src = await img.getAttribute('src')

      at it (integration/next-image-legacy/default/test/index.test.ts:1170:3)
      at runTests (integration/next-image-legacy/default/test/index.test.ts:1497:7)
      at integration/next-image-legacy/default/test/index.test.ts:1486:58
      at Object.describe (integration/next-image-legacy/default/test/index.test.ts:1485:1)

</details>


Read more about building and testing Next.js in [contributing.md](https://github.com/vercel/next.js/blob/canary/contributing.md).

**Pull request merged**